File: 544/f1.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
	
	#ifndef ONLINE_JUDGE
    	freopen("input", "r", stdin);
    	freopen("output", "w", stdout);
	#endif

    int n,m;
    cin>>n>>m;
    int countarr[n];
    vector<vector<int> > mainarr;

    for(int i=0;i<n;i++){
    	//mainarr[i]=(int*)malloc(sizeof(int));
    	vector<int> vec;
    	mainarr.push_back(vec);
    	countarr[i]=0;
    }

    int a,b;
    for(int i=0;i<m;i++){
    	cin>>a>>b;
    	a--;
    	b--;
    	mainarr[a].push_back(b);
    	countarr[a]++;
    	mainarr[b].push_back(a);
    	countarr[b]++;
    }


    int maxk=-1;
    int maxlen=0;

    for(int i=0;i<n;i++){
    	if(countarr[i]>maxlen){
    		maxlen=countarr[i];
    		maxk=i;
    	}
    }

    set<int> seenSet;
    queue<int> nextQueue;

    seenSet.insert(maxk);
    nextQueue.push(maxk);

    int nextval;
    int possnext;

    while(seenSet.size()<n && nextQueue.size()>0){
    	nextval=nextQueue.front();
    	nextQueue.pop();
    	for(int i=0;i<countarr[nextval];i++){
    		possnext=mainarr[nextval][i];
    		if(seenSet.find(possnext)!=seenSet.end()) continue;
    		cout<<nextval+1<<" "<<possnext+1<<endl;
    		seenSet.insert(possnext);
    		nextQueue.push(possnext);
    	}
    }
}

File: 544/f1.rb
require 'set'
if File.exists?("input")
  $stdin = File.open("input")
  $stdout = File.open("output","w")
end

nm = gets.chomp.split(" ").map(&:to_i)

n = nm[0]
m = nm[1]

hash = Hash.new(Array.new)

m.times do
	ab = gets.chomp.split(" ").map(&:to_i)
	a = ab[0]
	b = ab[1]

	hash[a]+=[b]
	hash[b]+=[a]
end

myset = Set.new

maxlen = 0
maxk=0
hash.each do |k,v|
	len = v.length
	if(len>maxlen)
		maxlen = len
		maxk=k
	end
end

myset.add(maxk)
nextqueue = Array.new

hash[maxk].each do |newnode|
	puts "#{maxk} #{newnode}"
	myset.add(newnode)
	nextqueue.push newnode
end

arr = []

while(nextqueue.size >0 && myset.size<n)
	nextval = nextqueue.shift
	hash[nextval].each do |possnext|
		next if myset.include?(possnext)
		arr.push [nextval,possnext]
		myset.add(possnext)
		nextqueue.push possnext
	end
end

arr.each do |x|
	puts "#{x[0]} #{x[1]}"
end